### PR TITLE
[DSv4.1] Add FROST prepared BF16 tail RoPE for prefill

### PR DESCRIPTION
**Pull request by YangXu1990uiuc**
Stacked on #1114. [Review the incremental diff](https://github.com/YangXu1990uiuc/cudnn-frontend/compare/14de2846757e8d452c06f0198f3d6e2aceec2991...b6bc4f725ef626552e4b3e7f5714d70acb74a706).
This PR targets `develop`, so its Files changed tab currently includes the
parent change. Rebase after #1114 merges; review the linked increment separately.

Add `TailRoPEForward` and `tail_rope` for contiguous BF16 `[T,D]` and `[T,H,D]`
with D=128/512 on SM100. The kernel copies the unrotated prefix and rotates the
last 64 channels using caller-prepared FP32 cosine/sine tables. Explicit FP32
multiply/FMA ordering preserves the source BF16 rounding and signed zeros.

The prepared `backend="frost"` API supports caller-owned output, dynamic token
counts, explicit streams, zero workspace and CUDA Graph capture. Execute performs
no allocation, conversion, synchronization or compilation. CuTe DSL >=4.7.0 is
required. Lazy exports, documentation, API tests and a standalone benchmark are
included. This change is stacked on the group32 RoPE/QDQ change.

Two independent B200 runs measured the actual public API against the fastest
valid measured FlashInfer, compiled-Torch or source control per case and metric:

| Prefill stage | Speedup across both runs and cache regimes |
| --- | ---: |
| GPU event latency | **1.04–1.69×** |
| Graph replay wall latency | **1.03–1.54×** |

Each run covers 64 cases: T=1/4/4096/16384; query, opposite-angle output,
index-query, KV and index-key geometries; TP1/TP8 local head counts; and source
base/compressed frequency tables. Activations are synthetic BF16. TP8 means
per-rank geometry, with no multi-rank execution or collectives. All prefill case
medians win on both metrics. Eight paired blocks and both cache regimes are
retained, including individual block win counts.

**Decode has regressions:** its GPU speedup spans 0.86–1.20×
and Graph wall spans 0.91–1.08×. The intended performance
use is prefill. Measurements include the full tensor copy, exclude table
preparation, and amortize three independent buffer pairs per replay. Graph wall
includes submission and completion waiting; it excludes eager API validation and
wrapper allocation. These are operator-stage measurements, with no complete
consumer, attention or full-model throughput claim.

Validation:

- All 32 new API tests and 29 parent QDQ tests pass on B200, including dynamic
  shapes, zero tokens, alias/alignment guards, explicit streams, changed-input
  Graph replay and a real tensor exceeding 2^31 elements.
- Four intentional faults are detected: device-to-host synchronization, ignored
  explicit stream, missing implicit-device binding and omitted Graph work.
- Both performance runs pass source, numerical, route, timing, GPU-isolation and
  thermal audits. Each records 3,456 numerical checks, 30,720 retained samples
  and 384 provider routes. Compiled PTX/CUBIN retention confirms all six public
  kernel specializations without changing execute.
- Frost matches the source bit for bit on seven input generations. Controls
  exact on both timed distributions remain competitors even when another stress
  generation fails; their witnesses and contract status remain recorded.
- CPU import/metadata checks pass, including declining old DSL before kernel
  import. The new kernel passes the naming guard; the pinned base has 17
  existing unrelated naming violations. The original GPU tests reused the
  pinned compiled FE binding; post-rebase host build validation is described
  below.

The API requires 16-byte-aligned, contiguous tensors and disjoint output. It has
no autograd rule and rejects gradients. Opposite-angle forward execution is not
a backward implementation. Quantization, table gathering and cache packing are
outside this change.

The arithmetic and frequency tables follow
[DeepSeek-V4.1-Flash](https://huggingface.co/deepseek-ai/DeepSeek-V4.1-Flash/blob/dba1be0a40aa45a94ad051997016db3960a90277/inference/model.py).
The vectorized kernel, API and benchmark glue are authored for this work. The
benchmark loads two unchanged source reference functions from a user-supplied,
hash-checked checkout. FlashInfer controls retain upstream attribution; the
no-FTZ control changes only a compiler flag on the same sources. Reproduction is
documented in `benchmark/rope_tail/README.md`.

Publication validation: rebased onto `develop` at
`adb963c01b6f65a0622efda48092fbc853dd3cf5`. Operation kernels, APIs,
benchmark controls and focused test files are byte-for-byte identical to the
validated versions described above. The rebase preserves the new upstream
lazy exports and documentation entries. Source identity, syntax, Black 26.3.1
(line length 160), SPDX headers and diff checks pass. Performance results above predate
this rebase.

Post-rebase host validation on these exact published heads now passes. A fresh
common C++ binding was built with GCC 11.4, CUDA 13.2.46, cuDNN 9.28.0 and
CPython 3.12.12; all binding sources, headers and CMake inputs are identical
across #1113–#1116. Each PR's byte-verified Python package passes actual import,
lazy-export and missing-dependency checks with that new binding. CPU metadata
checks also pass, with hardware queries explicitly mocked as SM100 and no GPU
kernel execution.

Fresh post-rebase B200 regression passed on 2026-09-17 at
`941ee629ed6a1d83a0ec876ad56ec73039768b7b`: **61 focused API tests passed**,
with zero failures, errors or skips, using the fresh binding and cuDNN 9.28.0.
All 165 focused tests across #1113–#1116 passed on the same B200; loaded
package files and the binding were checked against their source hashes.
This adds correctness and integration validation. Performance figures above
still refer to the original measured versions; this regression did not time kernels.

Compatibility follow-up: move the QDQ API from `rope/api.py` to `rope/qdq.py` so it can coexist with the independent Vision RoPE backward API in #1122. The moved implementation is byte-identical; public names, signatures and kernel modules are unchanged. Benchmark provenance now hashes the new API path. The group16 and tail branches carry the same move.

The updated Python packages pass real import, lazy-export and CPU metadata checks with a freshly built current-develop FE binding (cuDNN 9.26 headers/backend). A local integration preview of all six DSv4.1 PRs also verifies that Vision, QDQ and tail resolve to separate modules and that their compile helpers cannot overwrite one another. Hardware queries are explicitly mocked for these metadata checks; this follow-up adds no GPU execution or timing results.

The QDQ cache regression now imports `cudnn.rope.qdq`, so it observes the QDQ compiler rather than the Vision compiler in the combined package. This follow-up changes only that test import.

Combined GPU integration validation: a local merge preview of #1113, #1114, #1115, #1116, #1122 and #1179 passed **139 focused B200 tests with zero failures, skips or deselections** in one package: Engram 22, group32/group16 RoPE QDQ 53, BF16 tail RoPE 32, Vision backward 29, and mHC projection backward 3. This includes the >2^31-element tail case, changed-input graph replay, explicit streams, numerical references and the mHC two-kernel route check. The test process exited 0; all loaded FE files and the freshly built binding matched the frozen sources. Runtime versions were cuDNN 9.28, PyTorch 2.13/cu130, CuTe DSL 4.7, Triton 3.7.1 and cuda-tile 1.5. The binding was compiled from current-develop sources against cuDNN 9.26 headers.

The enclosing validation wrapper subsequently exited nonzero because its compiler-cache footprint exceeded a 512 MiB bookkeeping limit (about 637 MiB); that failure and the complete test artifacts are retained separately. This run adds combined-package correctness evidence, with no new performance measurements.

## Summary by CodeRabbit

- **New Features**
  - Added experimental SM100 RoPE APIs for BF16 tail rotation and in-place FP4/FP8 microscaled quantize-dequantize processing.
  - Added prepared execution support with CUDA streams, graph replay, dynamic lengths, and cached compilation.
  - Added strict validation for supported shapes, devices, layouts, alignment, memory overlap, and autograd usage.

- **Documentation**
  - Added API reference pages, navigation links, usage requirements, supported configurations, and benchmark methodology.

- **Tests**
  - Added comprehensive correctness, numerical precision, graph execution, and invalid-input coverage for both RoPE features.

### CODE REVIEW

File: benchmark/rope_tail/source_contract.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import json
+import math
+from pathlib import Path
+
+import torch
+
+
+def source_functions(source):
+    path = Path(source) / "model.py"
+    assert hashlib.sha256(path.read_bytes()).hexdigest() == "4e9ae23620edc8028ccc5d5fef552ab7fdc7dcd6f79608754fe9f67644056f65"
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply benchmark/rope_tail/source_contract.py]: Fixed in 4a6ed0021c65. Both digest gates now raise explicit `ValueError`, and parsing/compilation uses the same bytes that were hashed. CPU checks reject changed `model.py` and `config.json` under both normal Python and `python -O`. The pinned base and compressed frequency tables remain bit-identical in both modes. Kernel/API code is unchanged.

**Review by YangXu1990uiuc — COMMENTED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `b6bc4f725ef6` against base `9e9050c4013e`.

The tail RoPE increment looks sound: exact arithmetic ordering, dynamic Int64 indexing, validation and replay checks held up in the current-head run. The QDQ module relocation also preserves its implementation and now targets the correct compiler in the cache test. One nonblocking older-Torch stream compatibility comment remains. This review is comment-only because it uses the author's account.

**[P2] Normalize the raw default stream before allocating output** ([source](https://github.com/NVIDIA/cudnn-frontend/blob/b6bc4f725ef626552e4b3e7f5714d70acb74a706/python/cudnn/rope/tail.py#L164))

For stream=0 on Torch versions affected by the ExternalStream(0) bug, this context is a fresh pooled stream, while the later plan.execute still receives the original zero handle. Output allocation is therefore associated with a different stream from the kernel, invalidating the wrapper's allocation/launch ordering contract under asynchronous allocator reuse. Please map default-stream handles to torch.cuda.default_stream and use one resolved stream consistently; current develop already has cudnn._torch_stream.as_torch_stream for this (Python Hard Rule 5/R1). Add a raw-zero control alongside the existing nondefault-stream test. This is an older-runtime wrapper compatibility issue; the prepared path and tested Torch 2.13 path are unaffected.

Evidence: Using this exact wrapper on the cached Torch 2.11 runtime, with only TailRoPEForward replaced by a boundary-recording stub, stream=0 allocated output under a nonzero pooled stream but forwarded kernel_launch_stream=0. The actual default stream was zero. This probes ownership/dispatch, not numerical corruption or a tail-kernel execution on Torch 2.11. Independent current-head GPU numerics ran on Torch 2.13. Attribution: Introduced by the tail wrapper's direct ExternalStream construction; unchanged by the recent QDQ module relocation.

Validation: Reviewed the tail-specific increment over parent #1114 at 14de2846757e, including the API/kernel, tests, benchmark/source loader, exports and documentation. Also inspected the shared QDQ implementation and its integration after relocation. Exact-head local SM100 run with -m 'L0 or L1': 61 passed, zero skipped (32 tail RoPE plus 29 QDQ). CuTe DSL 4.7, Torch 2.13.0+cu130, cuDNN 9.26.0.51; verified imports from the review checkout and native-source compatibility of the binding. Includes BF16 boundary/signed-zero cases, dynamic/zero tokens, changed-input poisoned-output replay, explicit streams, rejection guards, and the real tensor exceeding 2^31 elements. Verified the tail API, kernel and focused tests are unchanged from 4a6ed0021c65. Relative to that head, the update is the byte-identical api.py -> qdq.py move plus four import/provenance-path adjustments; no new execute or kernel work was introduced. Independently confirmed both model.py and config.json hash gates reject changed bytes with optimize=0 and optimize=2. The earlier CodeRabbit hash-check issue is resolved. Current-head Style and Docs checks passed; git diff --check passed. No additional broad CI was launched during this review.

Limitations: The 64-case B200 performance claims remain author-provided evidence; this local SM100 correctness run does not independently verify those speedups. The latest relocation does not change the measured tail API/kernel, so no additional before/after timing was warranted for that delta. This is the review of the tail increment and shared-code integration; the parent #1114's separate benchmark/performance audit is not claimed here. The old-Torch stream probe substitutes the plan boundary and establishes differing allocation/launch streams; it does not claim an observed device-memory corruption.

Comment only: GitHub does not allow reviewing one's own PR with approval/request-changes.

**Comment by YangXu1990uiuc** [comment]
Addressed the raw-handle wrapper finding in 776c487fc. The wrapper now resolves the requested stream through `cudnn._torch_stream.as_torch_stream` and uses the same resolved handle for allocation and launch. The branch also incorporates develop so it uses the repository's shared resolver.

Added five L0 allocation/launch-boundary controls: raw 0, legacy/per-thread sentinels, current handle, and side handle. With the old wrapper, the three sentinel controls failed and the current/side controls passed; with the fix, all five pass. These controls simulate the older `ExternalStream` behavior and replace only the prepared plan boundary with a recording stub; they are not a claim of full kernel execution under older Torch.

The complete tail RoPE suite also passes: **37 passed, zero skipped**, with explicit L0/L1 selection on SM100, Torch 2.13/CUDA 13.0, cuDNN 9.28 and CuTe DSL 4.7. Imports and unchanged native binding inputs were verified against this checkout. The suite includes real kernel numerics, noncurrent-stream wrapper execution, changed-input Graph replay, and the Int64 extent case. Changed-file pre-commit checks pass. Prepared execute and device kernel code are unchanged; this is correctness validation, not a new performance measurement.

**Review by YangXu1990uiuc — COMMENTED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `c8b944edf60e` against base `9e9050c4013e`.

The previous raw-default-stream finding is resolved: allocation and launch now use the same shared-resolver result. Independently confirmed on the actual older Torch runtime that exposed the issue. The QDQ benchmark guards also now reject optimized Python before assertion-based validation can be disabled. No remaining findings in this follow-up. Comment-only because this PR belongs to the posting account.

Validation: Compared the 28 PR-specific files with previously reviewed b6bc4f725ef6: only the tail wrapper/test and five QDQ benchmark guard/documentation files differ. The prepared tail execute, device kernel and QDQ runtime implementation are unchanged; the large remaining head-to-head difference is incorporated develop history. Exact-head SM100 with Torch 2.13, CuTe DSL 4.7 and cuDNN 9.26: 66 passed, zero skipped, with explicit L0/L1 selection (37 tail RoPE and 29 QDQ). Includes real numerics, noncurrent streams, dynamic lengths, changed-input poisoned replay, the Int64-extent case and five new allocation/launch stream controls. Verified checkout imports and native-binding source compatibility. Actual Torch 2.11 boundary probe: requested raw handle 0, allocation stream 0, forwarded kernel-launch stream 0, canonical default 0. This uses a recording prepared-plan substitute, not a claim of full tail-kernel numerics on Torch 2.11; the old wrapper previously reproduced nonzero allocation versus zero launch. All four QDQ benchmark/source/auditor entry points reject both -O and -OO before imports or external-source loading: eight checks passed. Changed-file whitespace validation and current-head pre-commit/fern checks pass.

Limitations: Reused prior design/performance evidence for unchanged prepared/kernel code; no new speedup measurement. This follow-up does not turn the author's B200 performance results into an independent reproduction. No additional broad CI launched given the focused current-head validation and limited delta.

Comment only: GitHub does not allow reviewing one's own PR with approval/request-changes.

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run oss python_tests

**Review by vedaanta — APPROVED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `c8b944edf60e` against base `d40c621b7611`.

Reviewed the tail-RoPE increment after completing parent #1114. No actionable finding. The explicit multiply/FMA ordering, signed Int64 indexing, dynamic token-count ABI, caller-owned output, alias/layout checks and stream handling look sound. The previously reported raw-default-stream wrapper issue is resolved by using the shared resolver consistently for allocation and launch.

Validation: Exact-head L0/L1 execution on SM100, Torch 2.13+cu130, CuTe DSL 4.7 and cuDNN 9.26: 66 passed, zero skips (37 tail RoPE plus 29 parent QDQ). Verified the reviewed checkout is imported and reused native bindings only after byte-comparing their source inputs. Coverage includes bitwise BF16 arithmetic and signed zeros/subnormals, a negative control distinguishing the wrong FMA order, unchanged inputs, dynamic and zero token counts, execute allocation/compilation guards, changed-input graph replay after output poisoning, noncurrent streams, all five raw-stream allocation/launch controls, implicit-device binding, alias/layout/alignment rejection and a real tensor exceeding 2^31 elements. Read the twelve tail-specific changed files/integrations, benchmark/reference loader, docs, lazy exports and shared stream resolver. The other sixteen PR files are byte-identical to the independently reviewed parent d69d09e7. Changed-file whitespace check passed. Independent current-head prepared-execute spot check with bitwise numerical validation: CPU enqueue medians 19.3-20.1 us across five 100-call batches. Captured GPU times were 4.01/6.19 us for [4,4,128]/[4096,4,128], and 4.11/331.06 us for [4,64,512]/[4096,64,512]. These are local SM100 component measurements, not a comparison with FlashInfer or a reproduction of the author's B200 speedup.

Limitations: No fresh older-Torch runtime run; sentinel behavior was checked by the current-head emulation tests, and the prior reviewer separately reports an actual Torch 2.11 probe. No full B200 benchmark matrix, multi-rank or full-model validation. Existing CI is separate evidence and is not claimed wholly green; no duplicate broad pipeline launched.

Approved: no P0 or high-risk P1 found. Remaining findings stay with the owner; merge timing stays with the owner.

**Comment by YangXu1990uiuc** [comment]
Merged develop after #1114/#1115 landed. Tail API and kernel sources remain byte-identical to the reviewed head; the merged parent QDQ sources are preserved. Lazy exports, missing-dependency diagnostics, and targeted pre-commit hooks passed. This is an integration-only update, so the existing GPU pipeline and focused numerical validation remain the evidence; no full GPU rerun was requested.

The canceled SM107 OSS lane is outside this API's explicit SM100-only contract. All applicable GPU/build/analysis jobs passed; no SM107 coverage is claimed.
